### PR TITLE
Transcribe: support input media amr/flac/mp3/mp4/ogg/webm

### PR DESCRIPTION
**Pull request by janofsun**
Hi, here's my contribution for the "feature request: More input audio formats for Transcribe #6398" (#6818). Thank you for your time to review my code. If there is anything I need to modify, please let me know. 

Thank you for your support!

## Todo

- ~[ ] Add usage tracking for ffmpeg installer~ Current workaround is to treat the use of `transcribe` service as equivalent.

### CODE REVIEW

File: localstack/services/transcribe/provider.py
Comment (ackdav) [reply localstack/services/transcribe/provider.py:219]: Agreed - there needs to be a check whether conversion worked. Please refer to the docs in terms of Exceptions (e.g. [here](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/transcribe.html#TranscribeService.Client.start_transcription_job))

File: localstack/services/transcribe/provider.py
Comment (ackdav) [reply localstack/services/transcribe/provider.py:217]: Yes, agreed - it creates 2 output files, only 1 is needed. I quickly tried this out and while it actually works to create 2 files at once (here's how it would work cleanly: [link](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/transcribe.html#TranscribeService.Client.start_transcription_job)) - but again, no need to create 2 files

Comment (ackdav) [reply localstack/services/transcribe/provider.py:217]: Additionally, I don't think this ffmpeg command actually does what we need... L231 states that "PCM WAV" format is needed. [This link](https://stackoverflow.com/questions/4854513/can-ffmpeg-convert-audio-to-raw-pcm-if-so-how) suggests that a basic `ffmpeg -i` doesn't actually convert to pcm.

**Review by ackdav — CHANGES_REQUESTED** [review]
Dear @janofsun - thank you for your initiative! There are a couple of issues that would need to be addressed before using these changes (a couple are also mentioned [here](https://github.com/localstack/localstack/blob/master/CONTRIBUTING.md)):
- Please make sure to follow code formatting. Currently your PR is failing our linting check. (see link above)
- ffmpeg is currently not available within localstack - while this might work on your machine as a subprocess, localstack would need ffmpeg as a dependency. Additionally, ffmpeg is a sizeable dependency - isn't there a more lightweight solution? (maybe as a pip package)
- I would appreciate if you could add a few test cases ([here](https://github.com/localstack/localstack/blob/master/tests/integration/test_transcribe.py)) so we are certain your conversions actually work.  

Thanks!

**Comment by janofsun** [comment]
Dear @ackdav , thank you for your reply and helpful suggestions! I have already fixed almost those issues:
- I just redone 'make format' and 'make lint' to follow the code formatting.
- The apt package for ffmpeg is added. The pip package of ffmpeg seems to be no longer maintained, which is the reason why I choose apt method. And the 'amr' file is still unsupported for the transcription, which requires source code compilation.
- I have added tests for conversions in the test_transcribe.py file.

Thank you!

**Comment by janofsun** [comment]
Dear @mindriot101 , thank you for your reply and review! I have done couple of things to address those issues:
- Setting the check parameter as True to handle the instance failure.
- Changing the wav file path, then there is only one output file now.

Thank you!

**Comment by janofsun** [comment]
Dear @viren-nadkarni , thanks for your review! It is  true that ffmpeg is only used by the Transcribe provider, however, the unit-tests can not pass if ffmpeg is not installed during runtime. I would appreciate if you have other alternative methods to solve this. Thank you!

File: Dockerfile
Comment (baermat) [reply Dockerfile:52]: Yes, however we decided at the start of installer project that as long as it is not in master, dependencies should be added as they have been up until now. The overhead of organizing an adherence to the new model while it is still in the making is not worth it, better to treat it as yet another install routine that needs to be migrated afterwards :)

Comment (janofsun) [reply Dockerfile]: Hi, I have already moved the ffmpeg package installation process from the runtime stage to the base-full stage. Please feel free to check it out. Thank you!

File: localstack/services/transcribe/provider.py
Comment (ackdav) [inline localstack/services/transcribe/provider.py:233]: What if the conversion fails?
Scenario 1: User provides an invalid file
-> ffmpeg fails silently and no file is generated. `audio = wave.open(file_path, "rb")` will raise a `FileNotFoundError`.
this is ok since it's all in a try-except block, but could use a better `failure_reason`

Scenario 2: conversion process completes, but the generated file still does not have the right format.
-> the `failure_reason` would be "Audio file must be mono PCM WAV format", which is not helpful. Can you add a more helpful message to debug?

File: tests/integration/test_transcribe.py
Comment (ackdav) [inline tests/integration/test_transcribe.py:90]: Since you re-used check `is_transcription_done` from the function `test_transcribe_happy_path` - instead of having duplicate code, can we refactor this into a fixture?

**Review by ackdav — CHANGES_REQUESTED** [review]
Hi @janofsun 
Nice - always good to see some progress. A couple points to move forward:
- Regarding the test case: I highly encourage you to read the documentation here: https://docs.localstack.cloud/developer-guide/parity-testing/
Since your test case does not include a `snapshot.match("foo", bar)`, adding a snapshot at all, marking the test case with `aws_validated`, or `skip_snapshot_verify` would not be correct. I hope the documentation can make some of these points a bit more clear. 
- Regarding ffmpeg: I was a bit surprised to see the size being only a few MBs (https://packages.ubuntu.com/bionic/video/ffmpeg) - assuming dependencies don't add much more. @viren-nadkarni thoughts?
- Please see my comments inline regarding the code
Let me know if I should elaborate on any point. thanks!

File: localstack/services/transcribe/provider.py
Comment (viren-nadkarni) [reply localstack/services/transcribe/provider.py:222]: There's `ffprobe` which is bundled with `ffmpeg`. That will remove the dependency on `magic`

`ffprobe` can be used and parsed with something like:
```
$ ffprobe -show_format -print_format json -hide_banner -v error FILENAME
```

**Comment by viren-nadkarni** [comment]
> * Regarding ffmpeg: I was a bit surprised to see the size being only a few MBs (https://packages.ubuntu.com/bionic/video/ffmpeg) - assuming dependencies don't add much more. @viren-nadkarni thoughts?

@ackdav We can add ffmpeg to Dockerfile for now with a **TODO** that it must be refactored into a Transcribe dependency installer

File: Dockerfile
Comment (alexrashed) [inline Dockerfile:210]: Unfortunately, this is not an option (not even temporary). This drastically increases the Docker image size:
```
root@89c127bd5e65:/opt/code/localstack# apt install ffmpeg
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following additional packages will be installed:
  dbus fontconfig fontconfig-config fonts-dejavu-core i965-va-driver intel-media-va-driver libaacs0 libaom0 libapparmor1 libasound2 libasound2-data libass9 libasyncns0 libavc1394-0 libavcodec58 libavdevice58 libavfilter7 libavformat58 libavresample4 libavutil56
  libbdplus0 libbluray2 libbs2b0 libcaca0 libcairo2 libcdio-cdda2 libcdio-paranoia2 libcdio18 libchromaprint1 libcodec2-0.8.1 libcroco3 libcrystalhd3 libdatrie1 libdbus-1-3 libdc1394-22 libdrm-amdgpu1 libdrm-common libdrm-intel1 libdrm-nouveau2 libdrm-radeon1 libdrm2
  libelf1 libfftw3-double3 libflac8 libflite1 libfontconfig1 libfreetype6 libfribidi0 libgdk-pixbuf2.0-0 libgdk-pixbuf2.0-bin libgdk-pixbuf2.0-common libgl1 libgl1-mesa-dri libglapi-mesa libglib2.0-0 libglib2.0-data libglvnd0 libglx-mesa0 libglx0 libgme0 libgomp1
  libgraphite2-3 libgsm1 libharfbuzz0b libice6 libiec61883-0 libigdgmm5 libjack-jackd2-0 libjbig0 libjpeg62-turbo liblilv-0-0 libmp3lame0 libmpg123-0 libmysofa0 libnorm1 libnuma1 libogg0 libopenal-data libopenal1 libopenjp2-7 libopenmpt0 libopus0 libpango-1.0-0
  libpangocairo-1.0-0 libpangoft2-1.0-0 libpciaccess0 libpgm-5.2-0 libpixman-1-0 libpng16-16 libpostproc55 libpulse0 libraw1394-11 librsvg2-2 librsvg2-common librubberband2 libsamplerate0 libsdl2-2.0-0 libsensors-config libsensors5 libserd-0-0 libshine3 libslang2
  libsm6 libsnappy1v5 libsndfile1 libsndio7.0 libsodium23 libsord-0-0 libsoxr0 libspeex1 libsratom-0-0 libssh-gcrypt-4 libswresample3 libswscale5 libthai-data libthai0 libtheora0 libtiff5 libtwolame0 libusb-1.0-0 libva-drm2 libva-x11-2 libva2 libvdpau-va-gl1
  libvdpau1 libvidstab1.1 libvorbis0a libvorbisenc2 libvorbisfile3 libvpx5 libwavpack1 libwayland-client0 libwayland-cursor0 libwayland-egl1 libwebp6 libwebpmux3 libx11-6 libx11-data libx11-xcb1 libx264-155 libx265-165 libxau6 libxcb-dri2-0 libxcb-dri3-0 libxcb-glx0
  libxcb-present0 libxcb-render0 libxcb-shape0 libxcb-shm0 libxcb-sync1 libxcb-xfixes0 libxcb1 libxcursor1 libxdamage1 libxdmcp6 libxext6 libxfixes3 libxi6 libxinerama1 libxkbcommon0 libxrandr2 libxrender1 libxshmfence1 libxss1 libxtst6 libxv1 libxvidcore4
  libxxf86vm1 libzmq5 libzvbi-common libzvbi0 mesa-va-drivers mesa-vdpau-drivers shared-mime-info va-driver-all vdpau-driver-all x11-common xdg-user-dirs xkb-data
Suggested packages:
  default-dbus-session-bus | dbus-session-bus ffmpeg-doc i965-va-driver-shaders libasound2-plugins alsa-utils libbluray-bdj firmware-crystalhd libfftw3-bin libfftw3-dev jackd2 libportaudio2 opus-tools pciutils pulseaudio libraw1394-doc librsvg2-bin lm-sensors serdi
  sndiod sordi speex nvidia-vdpau-driver nvidia-legacy-340xx-vdpau-driver nvidia-legacy-304xx-vdpau-driver
The following NEW packages will be installed:
  dbus ffmpeg fontconfig fontconfig-config fonts-dejavu-core i965-va-driver intel-media-va-driver libaacs0 libaom0 libapparmor1 libasound2 libasound2-data libass9 libasyncns0 libavc1394-0 libavcodec58 libavdevice58 libavfilter7 libavformat58 libavresample4
  libavutil56 libbdplus0 libbluray2 libbs2b0 libcaca0 libcairo2 libcdio-cdda2 libcdio-paranoia2 libcdio18 libchromaprint1 libcodec2-0.8.1 libcroco3 libcrystalhd3 libdatrie1 libdbus-1-3 libdc1394-22 libdrm-amdgpu1 libdrm-common libdrm-intel1 libdrm-nouveau2
  libdrm-radeon1 libdrm2 libelf1 libfftw3-double3 libflac8 libflite1 libfontconfig1 libfreetype6 libfribidi0 libgdk-pixbuf2.0-0 libgdk-pixbuf2.0-bin libgdk-pixbuf2.0-common libgl1 libgl1-mesa-dri libglapi-mesa libglib2.0-0 libglib2.0-data libglvnd0 libglx-mesa0
  libglx0 libgme0 libgomp1 libgraphite2-3 libgsm1 libharfbuzz0b libice6 libiec61883-0 libigdgmm5 libjack-jackd2-0 libjbig0 libjpeg62-turbo liblilv-0-0 libmp3lame0 libmpg123-0 libmysofa0 libnorm1 libnuma1 libogg0 libopenal-data libopenal1 libopenjp2-7 libopenmpt0
  libopus0 libpango-1.0-0 libpangocairo-1.0-0 libpangoft2-1.0-0 libpciaccess0 libpgm-5.2-0 libpixman-1-0 libpng16-16 libpostproc55 libpulse0 libraw1394-11 librsvg2-2 librsvg2-common librubberband2 libsamplerate0 libsdl2-2.0-0 libsensors-config libsensors5 libserd-0-0
  libshine3 libslang2 libsm6 libsnappy1v5 libsndfile1 libsndio7.0 libsodium23 libsord-0-0 libsoxr0 libspeex1 libsratom-0-0 libssh-gcrypt-4 libswresample3 libswscale5 libthai-data libthai0 libtheora0 libtiff5 libtwolame0 libusb-1.0-0 libva-drm2 libva-x11-2 libva2
  libvdpau-va-gl1 libvdpau1 libvidstab1.1 libvorbis0a libvorbisenc2 libvorbisfile3 libvpx5 libwavpack1 libwayland-client0 libwayland-cursor0 libwayland-egl1 libwebp6 libwebpmux3 libx11-6 libx11-data libx11-xcb1 libx264-155 libx265-165 libxau6 libxcb-dri2-0
  libxcb-dri3-0 libxcb-glx0 libxcb-present0 libxcb-render0 libxcb-shape0 libxcb-shm0 libxcb-sync1 libxcb-xfixes0 libxcb1 libxcursor1 libxdamage1 libxdmcp6 libxext6 libxfixes3 libxi6 libxinerama1 libxkbcommon0 libxrandr2 libxrender1 libxshmfence1 libxss1 libxtst6
  libxv1 libxvidcore4 libxxf86vm1 libzmq5 libzvbi-common libzvbi0 mesa-va-drivers mesa-vdpau-drivers shared-mime-info va-driver-all vdpau-driver-all x11-common xdg-user-dirs xkb-data
0 upgraded, 180 newly installed, 0 to remove and 0 not upgraded.
Need to get 65.5 MB of archives.
After this operation, 367 MB of additional disk space will be used.
```

**Review by alexrashed — CHANGES_REQUESTED** [review]
Hey there!
I'm sorry that I have to chime in here, but imho adding `ffmpeg` to the docker image by default is not an option due to the sheer size of the dependencies (367 MB uncompressed).

File: Dockerfile
Comment (alexrashed) [reply Dockerfile:210]: Also, apt leaves quite some cache lying around after an update. It is best practice to remove the cache in the same layer / command (see above in the Dockerfile).

Comment (janofsun) [reply Dockerfile:210]: Hi Alexander, I did a comparsion: the sizes are 2.37G vs 2.23G. Is that 0.14G acceptable or do you have other alternative suggestion？

Comment (alexrashed) [reply Dockerfile:210]: Could you please share the method how you came to these values?
I guess it's two `make docker-build` with the size extracted from `docker image ls`?
There are some things to consider when talking about the docker image size:
- The interesting comparison is the "light" image, because it's the one which is pushed to `localstack/localstack`.
  - It can be built with `make docker-build-light`.
  - For me, the analysis looks as follows:
    ```
    # Build the light image without ffmpeg (without the changes on the PR)
    make docker-build-light
    docker tag localstack/localstack-light localstack/localstack-light-wo-ffmpeg
    # Get size of build with ffmpeg in bytes
    docker image inspect localstack/localstack-light-wo-ffmpeg | jq .[0].Size
    > 1864664703
    # Add the ffmpeg to the apt install / apply changes in the PR
    make docker-build-light
    # Get size of build with ffmpeg in bytes
    docker image inspect localstack/localstack-light | jq .[0].Size
    > 1993386742
    ```
  - So the addition of ffmpeg results in an increase of the uncompressed image size of 128722039 bytes or 122,76 MB.
  - (My results are a bit lower than yours because I added the ffmpeg install call to the apt install call in the base stage, where the cache is removed afterwards. Your `apt` call doesn't remove the apt cache.)
- The most interesting metric is the compressed image size (the one shown on DockerHub / the amount of data that needs to be downloaded).
  - I don't know of a way to reliably determine the compressed image size without pushing it to a registry.

While the 123 MB (uncompressed) seem like not too much, it's still an increase of 7% in image size just for this change. Since we have this use-case a lot, we try to tackle these problems by downloading the necessary software on demand (and cache it in the `var_libs` folder). In order to do that, we would need to have a statically linked ffmpeg build (assuming there are no issues with the license).
Do you know if there is a public downloadable statically linked build of ffmpeg for linux/arm64 and linux/amd64?

Comment (janofsun) [reply Dockerfile:210]: Yes, my comparing process is pretty much like yours. I got your idea and I will try to find a statically liked build of ffmpeg. Thanks!

Comment (janofsun) [reply Dockerfile:210]: I found this:
`https://github.com/jrottenberg/ffmpeg/blob/main/docker-images/5.1/ubuntu1804/Dockerfile`
Their ubuntu ffmpeg images is about 102MB which could potentially save us some space we compile the ffmpeg just like them. But there are two problems. 1. It doesn't support arm64. 2. It would drastically decrease the readability of our dockerfile. How do you think about it

Comment (alexrashed) [reply Dockerfile:210]: This would still add it directly to the docker image, instead of downloading a statically linked binary on demand when the transcribe service is used.
The best solution would be to find a publicly available download for both archs, which can be downloaded using a `PackageInstaller`, similar to the download of `kinesis-mock` here: https://github.com/localstack/localstack/blob/231547f3ce14e6d6ca2a5787f4204840c33383c7/localstack/services/kinesis/packages.py#L26

Comment (janofsun) [reply Dockerfile]: Hi Alexander, I found compiled builds of [ffmpeg](https://twitter.com/ffmpeg/status/1040386019722440704). And I changed the code based on your suggestion. Right now, we don't need to worry about the docker image size. There is a function which can download ffmpeg binary on the fly.

Comment (janofsun) [reply Dockerfile]: Hi Alexander, do you know how to solve that CLA issue? I am using a different computer. That bot seems recognize me  as other people.

**Comment by viren-nadkarni** [comment]
Hi @janofsun, firstly our apologies that this has stayed open for so long. We thank you for your contribution and patience. With respect to ffmpeg, we're currently discussing what its use will entail from a legal point of view. Once that is sorted, we can proceed towards merging this. 

In the meantime, could you perhaps give me push access to this branch so that I can contribute and help ready this PR?

**Comment by janofsun** [comment]
> Hi @janofsun, firstly our apologies that this has stayed open for so long. We thank you for your contribution and patience. With respect to ffmpeg, we're currently discussing what its use will entail from a legal point of view. Once that is sorted, we can proceed towards merging this.
> 
> In the meantime, could you perhaps give me push access to this branch so that I can contribute and help ready this PR?

Hi @viren-nadkarni, I invited you as collaborator to my forked repo. I believe you can contribute after accept. Thank you very much for your help!

File: localstack/services/transcribe/provider.py
Comment (viren-nadkarni) [reply localstack/services/transcribe/provider.py:249]: Leaving this out as a future `TODO`, based on how much this the transcribe service is used and its value.

**Review by alexrashed — APPROVED** [review]
Great to see this contribution getting over the line! This looks great now!
Thanks @janofsun for the contribution, and thanks @viren-nadkarni for polishing!
The only comment is a very minor nit (definitely not blocking).

File: localstack/constants.py
Comment (viren-nadkarni) [reply localstack/constants.py]: Fixed with a937bf04 :+1:

**Review by ackdav — APPROVED** [review]
Awesome, I added 2 nits around the tests, but overall LGTM! Thanks for bringing this over the line 🚀

**Pull request merged**
